# Social Computing: Notebook 1

**Deadline:** 18th March 2026, 12:00

### Preliminary information:
If you copy another group’s code in Jupyter Notebooks or give someone
your code to copy, both pairs involved in this behavior will get 0
points for the Notebook. This doesn’t mean you can’t consult someone
from another pair in case you and your coding partner are stuck on a
particular part of code though! You can do it, but you need to disclose
this in the Notebook. You can write something like this in a comment
next to the part of code on which you consulted someone - “we both
couldn’t find a solution for this on our own, so we asked students X and
Y for help. They explained us how this works. Now we see the issue was
that we were doing A instead of B”. Describe what prevented you from
figuring out stuff on your own to demonstrate you understood what’s
going on, not just copied code from another pair. If you provide a
proper explanation, you won’t lose the points for this.

#### Deliverables:

1. This notebook with your code and explanations/answers
2. Data collected: HTML pages and parsed data (JSON, CSV files, etc.) you used for analysis

- Please make sure to remove any code that you’ve used for testing and are not part of your solution, but keep the output cells.
- Rename this notebook file as described below.

**&rArr; You both have to submit your solution on OLAT, but you should submit the same files(s).**

#### How will you be graded?

You will be graded out of 100 points. Bonus tasks can earn you extra points, which you can use to make up for points lost in other notebooks.

---

Please include your names below and edit the name of the file to include your names (e.g. SC_Notebook1_Hamilton_Turing.ipynb):
- Alina Vanessa Brüllhardt
- Elias Mithanios Iskander

---

## 1) Online Data Collection Using the Requests Library (20 points)

### a) Familiarize yourself with HTML and requests (4 points)
#### HTML format
**Preliminary task:**
1. As a first step, pick your favorite Wikipedia page, open it in the browser, and then save it as an html file.
2. Now open it in the browser as well as in a text editor and look at the difference. What do you observe?

Your observations:
Saving the Page as an HTML removes all styling (CSS).

If I open it in the browser, the website stays fully functional. All styling is removed, but text links still work, texts are still defined as H1, H2, p and a. Buttons still show up and work. Basically, it all just reverts to their non styled, basic versions. All Spacing (Padding and Margins) are also removed

If I open it in a text editor, I get the "raw" HTML. The opened file does not at all resemble the original website, and doesn't retain any user functionality. I see all the different HTML Elements, as well as their attributes (like Class, Href, ID). This way, I can see how the website is built up. Basically, this is what I would see if I pressed "inspect" via my browser tools

#### Using Requests
Using the `requests` library, you can retrieve the html source of the page with the command `requests.get(url)`. The response object you receive has content that you can access calling the `text` function on it.

Write a script that **automatically** performs the following tasks:
1. Retrieve the Wikipedia page from the previous step using the `request` command.
2. Call `text` on the response object and save the result in a file.


Verify your solution:
- Open the file in a browser and check whether you successfully saved the page. Note: you will only be able to open the file in the browser if you give it an `.html` extension.

&rArr; Include the file in your submission.

---

**Hint**: Use the `headers` argument when making your `requests.get` call, otherwise it is likely that many websites reject your visit.

You can pass the following dictionary as an argument to the function call:

```python
headers = {'User-Agent': "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/109.0"}

requests.get(url=your_url, headers=headers)
```

---

In [ ]:
import requests
import os

# directory for HTML files (to keep our work space clean)
output_dir = "HTML Files 1a"
os.makedirs(output_dir, exist_ok=True)

wikipedia_page = requests.get(
    url="https://en.wikipedia.org/wiki/List_of_lists_of_lists", 
    headers= {'User-Agent': "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Mobile Safari/537.36"}
    )
result = wikipedia_page.text

print(result)
with open(f"{output_dir}/result_wikipedia.html", "w", encoding="utf-8") as f:
    f.write(wikipedia_page.text)

<!DOCTYPE html>
<html class="client-nojs skin-theme-clientpref-day mf-expand-sections-clientpref-0 mf-font-size-clientpref-small mw-mf-amc-clientpref-0 wp25eastereggs-enable-clientpref-1 skin-theme-clientpref-thumb-standard" lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<title>List of lists of lists - Wikipedia</title>
<script>(function(){var className="client-js skin-theme-clientpref-day mf-expand-sections-clientpref-0 mf-font-size-clientpref-small mw-mf-amc-clientpref-0 wp25eastereggs-enable-clientpref-1 skin-theme-clientpref-thumb-standard";var cookie=document.cookie.match(/(?:^|; )enwikimwclientpreferences=([^;]+)/);if(cookie){cookie[1].split('%2C').forEach(function(pref){className=className.replace(new RegExp('(^| )'+pref.replace(/-clientpref-\w+$|[^\w-]+/g,'')+'-clientpref-\\w+( |$)'),'$1'+pref+'$2');});}document.documentElement.className=className;}());RLCONF={"wgBreakFrames":false,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dm

### b) Basic web crawling (3 points)

URLs have specific formats, for example any Wikipedia page will be of the format https://en.wikipedia.org/wiki/Pythonidae where the last word is the topic of the article.
Next, we want to automate this saving process using the `requests` library and making automated requests to Wikipedia.

**Tasks:**
- Pick 5 different words and write code that loops through these words, retrieves the html content for each associated wikipedia page, and saves the html text as `wiki_htmls/[word].html` files (i.e. in a folder `wiki_htmls`). Include the folder with all files in your submission.
- Choose words that actually have associated wiki pages!


In [ ]:
import requests
import time
import os
# directory for HTML files (to keep our work space clean)
output_dir = "HTML Files 1b"
os.makedirs(output_dir, exist_ok=True)

words = ["List_of_lists_of_lists", "List_of_fictional_ducks", "List_of_animals_awarded_human_credentials", "Buttered_cat_paradox", "IP_over_Avian_Carriers"]

for word in words:
    wikipedia_page = requests.get(
    url= f"https://en.wikipedia.org/wiki/{word}", 
    headers= {'User-Agent': "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Mobile Safari/537.36"}
    )
    with open(f"{output_dir}/{word}.html", "w", encoding="utf-8") as f:
        f.write(wikipedia_page.text)
    time.sleep(2)

### c) URL formats (4 points)

#### Answer the following questions:

**1. What is the common URL (the part that stays the same) in the case of Google searches? And in the case of Yelp?**

Your answer: the domain. Google f.e has google.xxx/ and yelp has xx.yelp.xxx/
Yelp puts the language the site shows ahead of their domain (e.g de.yelp.ch/)
Google simply redirects to google.com, and the language selector is behind the .com (eg. google.com/de).

**2. And what happens to the URL if you want to define the location as well as the type of venue you are looking for?**

Your answer: 
In google, the google.com/ stays the same. It adds the query "/search?q=", and then adds the source youre form (f.e Homepage (google.com)), session id, search query, and some more user info.
In Yelp, the de.yelp.ch/ stays the same. It adds the query "/search?find_desc:", and then adds first your search term, and then your location.

**3. Can you find more search parameters for either of the two pages that you can define via the URLs? What do they mean?**

Your answer:
In google, you can add things like "" in the search (forcing the exact wording). This adds %22 at the beginning and end of the search query. In yelp, you can filter your search, which adds attrs (attributes) after your search query.

### d) Familiarize yourself with request and response headers (5 points)


Every request you send has a so-called HTTP header (unrelated to the content of the message), for example to communicate the size of the message, the browser from which the request is coming from, or what kind of response it is expecting back in the response.

1) Read up on this: What parts does a request contain exactly and what is the purpose of a header?

2) Look in the browser: Take a URL and find the request header using the developer tools in your browser. (Hint: you will need to look inside the “network” tab).

3) If you don’t tell python otherwise, it will use a default header when sending requests. What is this default when you use the requests library?

4) The requests library allows to specify the headers of your request exactly. Set the header of your request (for the  URL you previously picked) to be the one copied from your browser.

Your chosen URL: https://en.wikipedia.org/wiki/Main_Page

Default header of Python requests: 
{'User-Agent': 'python-requests/2.32.5', 'Accept-Encoding': 'gzip, deflate, zstd', 'Accept': '*/*', 'Connection': 'keep-alive'}

Request header in your browser (only the relevant ones):
User-Agent: Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Mobile Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate, br, zstd
Accept-Language: en-US,en;q=0.9
Cookie: GeoIP=CH:ZH:Zurich:47.39:8.52:v4; WMF-Uniq=8OdpUOV1h_sjDJzeBrzVfQIHACkKAFvdFU0K8HkkdheAezE8lEPfQ662T6RkNv5b; enwikimwuser-sessionId=38cc9f1b9d148976f1d4; WMF-Last-Access-Global=15-Mar-2026; WMF-Last-Access=15-Mar-2026; NetworkProbeLimit=0.001; WMF-DP=8f4,353,f19

In [ ]:
"""Default Python header"""
import requests
r = requests.get("https://en.wikipedia.org/wiki/Main_Page")
print(r.request.headers)

{'User-Agent': 'python-requests/2.32.5', 'Accept-Encoding': 'gzip, deflate, zstd', 'Accept': '*/*', 'Connection': 'keep-alive'}


5) Now compare the response headers for the same URL in the browser, and by calling a function on the response object in your code. What differences do you see?

In [ ]:
print(r.headers)

{'content-length': '126', 'content-type': 'text/plain', 'x-request-id': '98b76ff2-c72e-4cf6-b332-ca4f2b0db119', 'server': 'HAProxy', 'x-cache': 'cp6009 int', 'x-cache-status': 'int-tls', 'x-analytics': ''}


Response header in your browser:
{'accept-ch': '', 'accept-ranges': 'bytes', 'age': '7313', 'cache-control': 'private, s-maxage=0, max-age=0, must-revalidate, no-transform', 'content-encoding': 'gzip', 'content-language': 'en', 'content-length': '45680', 'content-type': 'text/html; charset=UTF-8', 'date': 'Sun, 15 Mar 2026 12:32:53 GMT', 'last-modified': 'Sun, 15 Mar 2026 12:32:45 GMT', 'server': 'mw-web.eqiad.main-8987d8bd9-qkgjg', 'strict-transport-security': 'max-age=106384710; includeSubDomains; preload', 'vary': 'Accept-Encoding,X-Subdomain,Cookie,Authorization,User-Agent', 'x-cache': 'cp6009 hit, cp6012 hit/2017', 'x-cache-status': 'hit-front', 'x-client-ip': '2a04:ee41:4:20f5:64cc:81ae:1e95:ce7d', 'x-content-type-options': 'nosniff', 'x-request-id': 'e2b39052-f1e9-4ed1-a9e8-186804338c2f'}

Response header in the response in Python:
{'content-length': '126', 'content-type': 'text/plain', 'x-request-id': 'a2213f98-6e90-4f18-bc7a-48f993a807df', 'server': 'HAProxy', 'x-cache': 'cp6012 int', 'x-cache-status': 'int-tls', 'x-analytics': ''}

Differences:
No Content Encoding in python -> Difference in Content Length
Server is different
Extra Headers in Browser that Python doesn't have (e.g nel, report-to, server-timing)
Browser shows my IP

### e) HTML content basics (4 points)

Let's examine a very simple webpage:

In [ ]:
# Run this cell
import requests
res = requests.get("http://dataquestio.github.io/web-scraping-pages/ids_and_classes.html")
# We do not need to include the header here

Using the **BeautifulSoup** parser library we will parse the webpage that you just saved.
To find out more about BeautifulSoup, visit: https://beautiful-soup-4.readthedocs.io/

In [ ]:
# let's import BeautifulSoup, our parser library
# And make a soup object out of the html of the page

from bs4 import BeautifulSoup
soup = BeautifulSoup(res.content, 'html.parser')

In [ ]:
# print a nice version using prettify
print(soup.prettify())

<html>
 <head>
  <title>
   A simple example page
  </title>
 </head>
 <body>
  <div>
   <p class="inner-text first-item" id="first">
    First paragraph.
   </p>
   <p class="inner-text">
    Second paragraph.
   </p>
  </div>
  <p class="outer-text first-item" id="second">
   <b>
    First outer paragraph.
   </b>
  </p>
  <p class="outer-text">
   <b>
    Second outer paragraph.
   </b>
  </p>
 </body>
</html>



Here's how we can find all instances of a tag at once.

1. Try to predict what the following command will return: `soup.find_all('p')`

Your prediction: It'll return all html paragraph (p) elements. Likely format <p class & extra stuff>Content</p>

2. Now call it to check if you were right.

In [ ]:
soup.find_all('p')


[<p class="inner-text first-item" id="first">
                 First paragraph.
             </p>,
 <p class="inner-text">
                 Second paragraph.
             </p>,
 <p class="outer-text first-item" id="second">
 <b>
                 First outer paragraph.
             </b>
 </p>,
 <p class="outer-text">
 <b>
                 Second outer paragraph.
             </b>
 </p>]

3. Print out the second element of this list.

In [ ]:
soup.find_all("p")[1]


<p class="inner-text">
                Second paragraph.
            </p>

4. Print out the text inside the second element of the list, using the `.text` on the element.

In [ ]:
soup.find_all("p")[1].text.strip()


'Second paragraph.'

When you try to find a specific element on a page you can reach it by finding classes or IDs of the elements.

In [ ]:
# Run this cell to see the output
soup.find_all('p', class_='outer-text')

[<p class="outer-text first-item" id="second">
 <b>
                 First outer paragraph.
             </b>
 </p>,
 <p class="outer-text">
 <b>
                 Second outer paragraph.
             </b>
 </p>]

5. How many elements would it return for `'inner-text'`? Guess, and check your guess by using the `find_all` command.

Your guess: 2 elements

In [ ]:
### Verify your guess by using the find_all command:
soup.find_all('p', class_='inner-text')

[<p class="inner-text first-item" id="first">
                 First paragraph.
             </p>,
 <p class="inner-text">
                 Second paragraph.
             </p>]

---

## 2) Finding Elements in the Browser (30 points)
### a) Weather forecasts (15 points)

Since every web page is different and html can get very large and messy, the easiest way to find elements that you are interested in is to start from the browser window. So next we will quickly look at how to find elements using the developer tools in your browser. Open the following webpage in your browser (preferably Chrome): http://forecast.weather.gov/MapClick.php?lat=21.3049&lon=-157.8579#.Wkwh8VQ-fVo

Find the **developer tools** in your browser. (In Chrome, it's View &rarr; Developer &rarr; Developer Tools or command+option+I on Mac and ctrl+option+I on Windows). You should end up with a panel at the bottom or the right side of the browser like what you see below. Make sure the **Elements** panel is highlighted.

When trying to find a specific element, you can right click on it on the page and select "Inspect". This will also open up the developer tools window. For example if we want to extract the current temperature value:

<img src="img/weather_1.jpeg" width="500"/>

<img src="img/weather_2.jpeg" width="800"/>


**Tasks:**
1. Retrieve the weather forecast for Honolulu by requesting the following url: https://forecast.weather.gov/MapClick.php?lon=-157.835&lat=21.286
2. Parse the html content of the page using `BeautifulSoup`. By using the find function, extract and print out the current temperature (in Farenheit) from the page.
3. Now do the same with the value in Celsius.


In [ ]:
import requests
from bs4 import BeautifulSoup

soup = BeautifulSoup(requests.get("https://forecast.weather.gov/MapClick.php?lon=-157.835&lat=21.286").text, "html.parser")

temp_F_data = soup.find("p", class_="myforecast-current-lrg")
temp_C_data = soup.find("p", class_="myforecast-current-sm")

temp_F_clean = temp_F_data.get_text(strip=True)
temp_C_clean = temp_C_data.get_text(strip=True)


### Fill out and print a full sentence describing the temperature in F and C.
temp_F = f"The temperature in Honolulu, HI is {temp_F_clean}"
temp_C = f"Which translates to {temp_C_clean}"

print(temp_F)
print(temp_C)


The temperature in Honolulu, HI is 76°F
Which translates to 24°C


4. Now we want to extract each half day's forecast from the extended forecast on the weather report page. <br>
    a. Find the container for the extended forecast on the weather page we just downloaded. <br>
    b. Make a list with all forecast items (overnight, Wednesday, Wednesday night, etc.) <br>
    c. For each time period, print out the name of the period, the short description of the expected weather conditions, and the temperature. Try to output a nicely formated string.

**Example output for one time period:** <br>
_Overnight, the weather will be mostly clear and breezy and the temperature will be 65F._

In [ ]:
import requests
from bs4 import BeautifulSoup

soup = BeautifulSoup(requests.get("https://forecast.weather.gov/MapClick.php?lon=-157.835&lat=21.286").text, "html.parser")

ul = soup.find("ul", class_="list-unstyled")
for li in ul.find_all("li"):
    period = li.find("p", class_="period-name").get_text(separator=" ",strip=True)
    temp = " ".join(li.find("p", class_="temp").get_text(strip=True).split(":")[1].split())
    desc = li.find("p", class_="short-desc").get_text(separator=" ", strip=True)

    print(f"{period}, the weather will be {desc} and the temperature will be {temp}")



Today, the weather will be Showers and Windy and the temperature will be 82 °F
Tonight, the weather will be Showers and Breezy then Scattered Showers and the temperature will be 71 °F
Monday, the weather will be Sunny then Scattered Showers and the temperature will be 81 °F
Monday Night, the weather will be Showers and the temperature will be 70 °F
Tuesday, the weather will be Showers Likely and the temperature will be 78 °F
Tuesday Night, the weather will be Scattered Showers and the temperature will be 68 °F
Wednesday, the weather will be Isolated Showers and the temperature will be 81 °F
Wednesday Night, the weather will be Showers Likely and the temperature will be 70 °F
Thursday, the weather will be Scattered Showers and the temperature will be 81 °F


### b) JobScout24 (15 points)

**Tasks:**
- Take a list of 5 jobs (e.g.['teacher', 'lawyer', 'data-scientist', …]).
- Write a script that **automatically** performs the following tasks:
    1. For each job in your list, save the html of the result of searching it on [JobScout24](https://www.jobscout24.ch/en/). (Hint: Examine the url of a result page.)
    2. For each job, find the names of the companies from the first result page. Create a dictionary where the keys are the jobs and the values are lists of the company names for the corresponding job.
    3. Print the dictionary.
    4. Additionally, store the result as a json file.

**Hint**: Use the `headers` argument in your requests for this exercise.

#comment: Teacher is often employed by a canton (e.g. Nidwalden), this is not the location but the employer.

In [ ]:
import json
import os

# directory for HTML and json files (to keep our work space clean)
output_dir = "HTML and JSON Files 2b"
os.makedirs(output_dir, exist_ok=True)

five_jobs = ['teacher', 'lawyer', 'scientist', 'banker', 'psychologist']
job_dict = {}

for job in five_jobs:
    #initiate dictionary entry with for Job as key
    job_dict[job]= []
    #1. get the html files
    jobscout_page = requests.get(
    url= f"https://www.jobscout24.ch/de/jobs/{job}", 
    headers= {'User-Agent': "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Mobile Safari/537.36"}
    )
    with open(f"{output_dir}/{job}.html", "w", encoding="utf-8") as f:
        f.write(jobscout_page.text)
    soup = BeautifulSoup(jobscout_page.content, 'html.parser')
    #2. Get company names
    companies = soup.find_all('span', class_='company')
    for company in companies:
        name = company.get_text(strip=True)
        #append companies to our job dictionary
        job_dict[job].append(name)

    time.sleep(2)

#3. save json
json_str = json.dumps(job_dict, indent=4)
with open(f"{output_dir}/jobs.json", "w") as f:
    f.write(json_str)

#4. print dictionary
print(job_dict)

{'teacher': ['Kanton Nidwalden', 'Kanton Nidwalden', 'Suva', 'Universitätsklinik Balgrist', 'GES School SA', 'Quantinum Lernwelten GmbH', 'Christophorus-Schule Basel', 'Kanton Solothurn', 'ipso Bildung AG', 'Gemeinde Schleitheim', 'Neue Stadtschulen AG', 'PureGym AG', 'Bischöfliches Ordinariat Der Diözese Basel', 'Jobcom GmbH', 'Kanton Obwalden', 'SWS Schulen für Wirtschaft und Sprachen Winterthur AG', 'Berner Bildungszentrum Pflege AG', 'Chiffre', 'Berner Bildungszentrum Pflege AG', 'Sonnegg', 'Kreisschule aargauSüd', 'Primarschule St. Stephan', 'Schweizerische Alpine Mittelschule Davos (SAMD)', 'SIU - Schweizerisches Institut für Unternehmerschulung AG'], 'lawyer': ['Anwaltskanzlei Grischott-Domanig AG', 'VISCHER AG', 'Reetz Sohm Rechtsanwälte', 'Wehrli Partner Rechtsanwälte AG', 'ZUMWALD & PARTNER AG', 'Centre Patronal', 'OBT AG', 'Gebäudeversicherung Bern - GVB Gruppe', 'Wieduwilt Rechtsanwälte AG', 'Integritas Compliance AG', 'Orion Rechtsschutz-Versicherung AG', 'Eidgenössische F

---

## 3) ImmoScout24 (50 points + 10 bonus points*)
By now, you should be ready to solve the following exercise without detailed step-by-step instructions.

- Navigate to the [ImmoScout24](https://www.immoscout24.ch) website.
- Try out different filters and observe how the URL changes.
- Choose a postal code of your choice, along with some other filters like radius (in km), number of rooms etc.

Now, write a script that **automatically** performs the following tasks:
1. Request the results page for your choice of postal code and filters.
2. Visit every result item (apartment)'s page and scrape some info. Please ensure you have information about at least 15 apartments. This info can be information about facilities in the vicinity, if pets are allowed or not, etc.
3. Collect this information for all items and store the combined data in a suitable format, i.e. a JSON or CSV file. Don't forget to include this file in your submission.
4. Have a conclusion at the end, e.g.: _In the neighbourhood of postal code 800X, X% of apartments are within 5 mins to a school. Whereas, in the postal code 800Y, only Y% of apartments are within 5 mins of a school._


#### Bonus Task:
Apply a set of filters that returns more than 50 results – i.e. at least 3 results pages – and use pagination to navigate across them and collect information about all the individual apartments.

\* _This bonus task can earn you up to 10 extra points._

---

#### Important Hints:

##### 1) Request Headers

You need to include several fields in your request header, especially a valid session cookie. Otherwise, you will only retrieve empty html pages with an error message. This might be different for each case: For some websites you only need to specify a user agent or no header at all, while for others, additional parameters are required.

To learn more about headers and their importance for web scraping, the following source might be helpful: https://brightdata.com/blog/web-data/http-headers-for-web-scraping

**Including the following header parameters should do the trick for ImmoScout24:**
- user-Agent
- accept
- accept-encoding
- accept-language
- cache-control
- cookie  (&rarr; this one is especially important!)

```python
# Replace [...] with the actual values that you've obtained using the developer tools in your browser.
headers = {
    'User-Agent': "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Mobile Safari/537.36",
    'accept': "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
    'accept-encoding': "gzip, deflate, br, zstd",
    'accept-language': "en-US,en;q=0.9",
    'cache-control': "max-age=0",
    'cookie': """homegate_language=DE; SUUID=6e731f18-af5c-44f9-a51a-9bbde6ccd443; OptanonAlertBoxClosed=2025-11-04T01:35:31.086Z; eupubconsent-v2=CQaXAdgQaXAdgAcABBDECDFsAP_gAEPgAAQ4LstR_G__bWFr-Tb3aftkeYxP99hj6sQhBgaBk2QFyDuS7JQWx2ExFAyKpiICARIAi3RBIQNlHBBERVCgCIgFJQDMYESUgBEKIABkiEERAkJICAwOiIkBWQCI5Pp9tlU1mBqN7ZLsWMzgy4gHj3a5y2QBEJCQAYMNDPh4YBKS85IEd-x4qwtwsFrhAGuOCRnfgGshqB8uYlMZZBEQt-bSffzPn__rl_e7X_vd_n37v94XH77v____f_-7___2C7QAJhoVEERZEAAQCBhBAgAUFYQAUCAIAAEgaIAAEgYEOQEAFVhMgBACgAGAAEAAIMAAQAACQAIRABQAACAACAQKAAEAAAICABAYAAgAUIAAAAIDoCIAEEAgWACRiVQaYEoACQQAlhAgEAQIK4QgEngAACIkAgAABAAKAAAAOCwEJJASsACALCCYAAAgAAAiBAgBSBGAIKAyQYAECTqMjTAMHzBMkp0GQBMEZGSbEJvwmHikKIUEOQGxSzAHQAAA.f_wACHwAAAAA.ILtNR_G__bXFv-Tb3aftkeYxf99hj6sQhBgaBs2QFyDuS7JQW12EzFEyKpiYCARIAi3RBIQNtHBhERVCgCIgFJRDsYESUgBkKIEBkiGERQkJICAwOiIkBWQCJ5Pp_t1V9mTqN7ZLuWMzgy5gnr3a5y-QBEJiQCYMNHPh4YBKS87IU9-x66wtwsNrhAGuOCRnfkGslqF8u4tPZZhEQt-7yffz____73_e7X__d_______Xf_7_____________AAA; AwinChannelCookie=direct; __rtbh.ssgtm.aid=wuzcQpGd4dWNdzBPkB4; homegate_language=DE; homegate_language=DE; alloy-init=eyJzbWciOiJleUp0WkNJNmV5SmtZWFJoSWpvaWUxd2lZMmwyWENJNmRISjFaU3hjSW5WcGNHaGhZMmxjSWpwbVlXeHpaU3hjSW1kaVkybGNJanBiWFN4Y0ltUnNZWFJsWENJNlczdGNJbXRsZVU1aGJXVmNJanBjSW1WMlpXNTBYQ0lzWENKclpYbFdZV3gxWlZ3aU9sd2lkbWx5ZEhWaGJGOXdZV2RsZG1sbGQxd2lMRndpWlhabGJuUlFZWFJvWENJNmJuVnNiSDFkTEZ3aVkybHlYQ0k2WENKY1hGeGNMeWhjWEZ4Y1pDc3BKRndpTEZ3aWMzVmpYQ0k2VzExOUlpd2laWGh3YVhKbGMwRjBJam94Tnpjek5qYzBOVEF5T0RrNWZYMD0ifQ==; alloy-cirsa=eyJzbWciOiJleUpoZGlJNmRISjFaU3dpYkdNaU9pSXlNREkyTFRBekxURTFWREUxT2pJeE9qUXlMamszTVZvaUxDSmhJam94TENKamRXa2lPaUppTXpkaVptTTFaalE0Wm1JMFltVTFPRFkwTlRFMVlqa3pZelptT1RGak1TSXNJbU4xYVdNaU9qRTNOek0xT0RneE1ESXNJbTUxYVdNaU9tNTFiR3dzSW5Sd1l5STZabUZzYzJVc0luVnBJanA3ZlgwPSJ9; _sp_ses.f306=*; g_state={"i_l":1,"i_ll":1773588102904,"i_b":"70SwjNDutbu4ynhDbSO5PxFXSRFGMFySR/qCXg+E3IY","i_e":{"enable_itp_optimization":0},"i_p":1773595304970}; __wk_eid=1z68652826151281M0888685188600188282L; __cf_bm=7cIdYGZF7vjKwFGFDy7WMxdyczjkQhvWUCOOMbCO_xI-1773588545-1.0.1.1-CRDbqEypRmgwS.Ms_uD1snEC6al9Vl0fth.d26dmB7zIigmYGvTz0pfYNKTTWJ_uWjBr_dQHlpB80CY93LotZRuASgiVJuiqdFmWpIZWt6k; _cfuvid=FJ0b._l8k3PX16_bEE.vov8M83A0Vw7lB2b7c9Tlr.c-1773588545881-0.0.1.1-604800000; cf_clearance=bJRUJKscLgHuh.17I_kT6GcPsHFzwRDPbvSBHB7g1fU-1773588546-1.2.1.1-qr48FgqtSMSyvFdmJ036VjLfTd69U7EfAhlg9TtQkvcZuWnQQA8h6VGPa5f4Ti8MnoBrJPyisubJg0rA.BgNQcTGpDEd86Pr.umlhcKVdJNtL5pmQrzvRGcyeAC7UUu5QM.CyXrQim4XnModdUlWPrBS834ZSYy_Atiyqp_ElHn9sT41SS4GhMLAKhUNOtbu6T_mTAc5p0NmJ5T35ePed1nr334prmiTTKgcwbh2saA; datadome=hkTXpzR7m6MKZ9v~AdS34pCdFzctm~iM3I84Wa85V1MPdrs2_p9kkHYH~Jo68KDrhYBvf9am83Qp3Avc~Sm7ucsbJtcleLe1YNMeTQ39zAYeh6ATlYD5D6f_6Jem5Rsi; OptanonConsent=isGpcEnabled=0&datestamp=Sun+Mar+15+2026+16%3A34%3A47+GMT%2B0100+(Central+European+Standard+Time)&version=202601.2.0&browserGpcFlag=0&isIABGlobal=false&hosts=&consentId=ceb5b641-36dc-42df-a9dd-e86be4b518d8&interactionCount=2&isAnonUser=1&landingPath=NotLandingPage&groups=C0003%3A1%2CC0001%3A1%2CC0004%3A1%2CC0002%3A1%2CV2STACK42%3A1&intType=1&geolocation=CH%3BZH&AwaitingReconsent=false&prevHadToken=0; _sp_id.f306=49cf68aa-22d3-47df-a95a-12eae73a39ca.1759679487.19.1773588887.1769192998.4aa10675-eb0c-46cf-ae37-aecebe193b1b.a82e0441-bb2a-44d5-8bfb-6e6f00661edd.add18a79-9577-4be5-a9ed-f750eb815dd5.1773588102999.4; _dd_s=aid=e4db6470-085e-4543-ab75-ccc3701d9a00&logs=0&expire=1773589957208&rum=0"""
}

requests.get(url=, headers=headers)
```

##### 2) Avoid Rate Limits

Add delays in your code to avoid being rate-limited.

---

In [14]:
### Tasks 1–3 and optional bonus task
import requests
import os
from bs4 import BeautifulSoup
import json
import time



output_dir = "HTML and CSV Files 3"
os.makedirs(output_dir, exist_ok=True)

############### 1. Request the results page for your choice of postal code and filters ###############
#https://www.immoscout24.ch/en/real-estate/rent/postcode-8008-zuerich?r=2000&pt=1400

postcode = 'postcode-8000'
city = '-zurich'
radius = '?r=2000'
rent = '&pt=1400'

# Combine them with "; "
cookie = """_cfuvid=JhzioItZ7maceA9Wy6fLGcEkk4MRshRo8TszC_BCBMk-1773529342565-0.0.1.1-604800000; homegate_language=EN; homegate_language=EN; JSESSIONIDCMS=C2838FEC15DB049FD6EE118FC3E21B5B; g_state={"i_l":0,"i_ll":1773595685032,"i_b":"I4MJRyzc2VexIQXON0A2UDzkv/aGktCxP545INItBOY","i_e":{"enable_itp_optimization":0}}; homegate_language=EN; __cf_bm=w3IiGz.9TyNaj54wjV6nRHgJl5zPTJLyHLULsqWSaCY-1773610810-1.0.1.1-NCqqfuAqq0fmIxwNf.8J1svFEzki1cUl2XmTwqte9zALWU6zVoKuapBaulTYgQ9rJ9xhaycIzM2lnBy17ny2qJqKjoXWxPplii2iq.yzHnA; cf_clearance=QFXSe4B_2Cz.FIkUxm8aH7IPIssqf3SdgSc64CffrAo-1773610811-1.2.1.1-N9srGAMDQTUOifciF1lWtwJ5Xfo6f2JEJ9pH9H5q1dlJtJo6_QmYuCm01q1RaA55cgWiojzn8TWjaBmqP5IBuYIJb6VWiDilravH0QZoN09iDSLjzomOdkXzSKFiE3DOGDLkGvLJ1RDer.Jb5cxJhU56MZG20qb3fpu0KDT6JUwVZkiNe0hhYg254Az8XsGKZlXsBaZuE.YB2HApaMQsq2Z2TXUZUCQ_QzkFLnq60rA; datadome=mfycQ2O8rjqvylaVpljGS5z_hUL6sOYKwKFaBK22_EKExd5UXjw9xl8tF482o41Am5coiVx3WafrAhX4cHWyJfb2~S0yNia5RS8h8bGCInq0lCDhF7~NT3r2Cp5k2~w~; __wk_eid=5551224246101421M5222620122641424444T; _dd_s=aid=f9b84a7c-416d-4fe0-bdcc-a77865885107&logs=0&expire=1773611965197&rum=2&id=36fcecd5-62cf-4597-83ac-394f74c46d1d&created=1773609546267"""

headers = {'User-Agent': "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Safari/537.36",
    'accept': "application/json, text/plain, */*",
    'accept-encoding': "gzip, deflate, br, zstd",
    'accept-language': "en-US,en;q=0.9",
    'cache-control': "max-age=14400, public",
    'cookie': cookie
    }       

# immoscout_page = requests.get(
#     url= "https://www.immoscout24.ch/en/real-estate/rent/postcode-8008-zuerich?r=2000&pt=1400", 
#     headers= headers
#     )

# with open(f"{output_dir}/immoscout_results.html", "w", encoding="utf-8") as f:
#         f.write(immoscout_page.text)

with open(f'{output_dir}/immoscout_results.html', 'r', encoding = 'utf-8') as f:
    immoscout_page = f.read()

############### 2. Visit every result item (apartment)'s page and scrape some info ###################
soup = BeautifulSoup(immoscout_page, 'html.parser')
a_class = soup.find_all('a', class_='HgCardElevated_content_900d9 HgCardElevated_link_6124c') 
#find the links of each apartment
links = [link.get('href') for link in a_class]

apartment_stats = {}
for link in links:
        apartment_stats[link] = {} #store info with key value pairs within the apartment_stats, where key is the link
        apartment_page = requests.get(
        url= link,
        headers= headers
        )
        apartment_soup = BeautifulSoup(apartment_page.text, 'html.parser')

        #for things like school, pharmacy, school: distance

        description = [desc.get_text(strip=True) for desc in apartment_soup.find_all('h4') 
                       if "Source of this data" not in desc]
        
     
        name = [names.get_text(strip=True) for names in apartment_soup.find_all('span', class_= "TravelTime_travelTimeListDescription_1c4d0")] 
        travel_time = [eta.get_text(strip=True) for eta in apartment_soup.find_all('span', class_= "TravelTime_travelTimeListTime_b30e4")]
        deets = zip(name, travel_time)

        apartment_stats[link].update(dict(zip(description, deets)))
        time.sleep(5)


############### 3. Collect this information for all items and store the combined data ################
json_str = json.dumps(apartment_stats, indent=4)
with open(f"{output_dir}/apartments.json", "w") as f:
    f.write(json_str)


**4. Your conclusion:**
…